From e48d32e5699d824dc229f131246229f077709f84 Mon Sep 17 00:00:00 2001
From: Jonathan Kang <jonathan121537@gmail.com>
Date: Tue, 19 Sep 2017 17:00:18 +0800
Subject: [PATCH 2/2] wifi: Add domain-suffix-match entry

domain-suffix-match has been supported in nm-connection-editor for a
while. So users can edit that property in nm-connection-editor. But
it's still not available in gnome-control-center yet.

Fix that by adding domain-suffix-match entry in tls, ttls and peap
security settings page.

https://bugzilla.gnome.org/show_bug.cgi?id=787882
---
 panels/network/wireless-security/eap-method-peap.c | 21 ++++++++++
 .../network/wireless-security/eap-method-peap.ui   | 43 ++++++++++++++++----
 panels/network/wireless-security/eap-method-tls.c  | 29 ++++++++++++-
 panels/network/wireless-security/eap-method-tls.ui | 47 +++++++++++++++++-----
 panels/network/wireless-security/eap-method-ttls.c | 21 ++++++++++
 .../network/wireless-security/eap-method-ttls.ui   | 39 +++++++++++++++---
 6 files changed, 175 insertions(+), 25 deletions(-)

Index: gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-peap.c
===================================================================
--- gnome-control-center-3.29.4.orig/panels/network/wireless-security/eap-method-peap.c	2018-07-17 02:32:35.000000000 +0200
+++ gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-peap.c	2018-07-24 12:14:46.053325821 +0200
@@ -114,6 +114,10 @@ add_to_size_group (EAPMethod *parent, Gt
 	g_assert (widget);
 	gtk_size_group_add_widget (group, widget);
 
+	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_domain_label"));
+	g_assert (widget);
+	gtk_size_group_add_widget (group, widget);
+
 	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_version_label"));
 	g_assert (widget);
 	gtk_size_group_add_widget (group, widget);
@@ -159,6 +163,12 @@ fill_connection (EAPMethod *parent, NMCo
 	if (text && strlen (text))
 		g_object_set (s_8021x, NM_SETTING_802_1X_ANONYMOUS_IDENTITY, text, NULL);
 
+	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_domain_entry"));
+	g_assert (widget);
+	text = gtk_entry_get_text (GTK_ENTRY (widget));
+	if (text && strlen (text))
+		g_object_set (s_8021x, NM_SETTING_802_1X_DOMAIN_SUFFIX_MATCH, text, NULL);
+
 	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_ca_cert_button"));
 	g_assert (widget);
 	filename = gtk_file_chooser_get_filename (GTK_FILE_CHOOSER (widget));
@@ -424,11 +434,22 @@ eap_method_peap_new (WirelessSecurity *w
 	                  (GCallback) wireless_security_changed_cb,
 	                  ws_parent);
 
+	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_domain_entry"));
+	if (s_8021x && nm_setting_802_1x_get_domain_suffix_match (s_8021x))
+		gtk_entry_set_text (GTK_ENTRY (widget), nm_setting_802_1x_get_domain_suffix_match (s_8021x));
+	g_signal_connect (G_OBJECT (widget), "changed",
+	                  (GCallback) wireless_security_changed_cb,
+	                  ws_parent);
+
 	if (secrets_only) {
 		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_anon_identity_label"));
 		gtk_widget_hide (widget);
 		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_anon_identity_entry"));
 		gtk_widget_hide (widget);
+		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_domain_label"));
+		gtk_widget_hide (widget);
+		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_domain_entry"));
+		gtk_widget_hide (widget);
 		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_ca_cert_label"));
 		gtk_widget_hide (widget);
 		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_peap_ca_cert_button"));
Index: gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-peap.ui
===================================================================
--- gnome-control-center-3.29.4.orig/panels/network/wireless-security/eap-method-peap.ui	2018-07-17 02:32:35.000000000 +0200
+++ gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-peap.ui	2018-07-24 12:14:46.053325821 +0200
@@ -67,6 +67,33 @@
           </packing>
         </child>
         <child>
+          <object class="GtkLabel" id="eap_peap_domain_label">
+            <property name="visible">True</property>
+            <property name="can_focus">False</property>
+            <property name="label" translatable="yes">_Domain:</property>
+            <property name="use_underline">True</property>
+            <property name="mnemonic_widget">eap_peap_domain_entry</property>
+            <property name="xalign">1</property>
+          </object>
+          <packing>
+            <property name="left_attach">0</property>
+            <property name="top_attach">1</property>
+          </packing>
+        </child>
+        <child>
+          <object class="GtkEntry" id="eap_peap_domain_entry">
+            <property name="visible">True</property>
+            <property name="can_focus">True</property>
+            <property name="hexpand">True</property>
+            <property name="activates_default">True</property>
+            <property name="tooltip_text" translatable="yes">Suffix of the server certificate name.</property>
+          </object>
+          <packing>
+            <property name="left_attach">1</property>
+            <property name="top_attach">1</property>
+          </packing>
+        </child>
+        <child>
           <object class="GtkLabel" id="eap_peap_ca_cert_label">
             <property name="visible">True</property>
             <property name="can_focus">False</property>
@@ -77,7 +104,7 @@
           </object>
           <packing>
             <property name="left_attach">0</property>
-            <property name="top_attach">1</property>
+            <property name="top_attach">2</property>
           </packing>
         </child>
         <child>
@@ -88,7 +115,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">1</property>
+            <property name="top_attach">2</property>
           </packing>
         </child>
         <child>
@@ -103,7 +130,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">2</property>
+            <property name="top_attach">3</property>
           </packing>
         </child>
         <child>
@@ -117,7 +144,7 @@
           </object>
           <packing>
             <property name="left_attach">0</property>
-            <property name="top_attach">3</property>
+            <property name="top_attach">4</property>
           </packing>
         </child>
         <child>
@@ -135,7 +162,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">3</property>
+            <property name="top_attach">4</property>
           </packing>
         </child>
         <child>
@@ -149,7 +176,7 @@
           </object>
           <packing>
             <property name="left_attach">0</property>
-            <property name="top_attach">4</property>
+            <property name="top_attach">5</property>
           </packing>
         </child>
         <child>
@@ -167,7 +194,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">4</property>
+            <property name="top_attach">5</property>
           </packing>
         </child>
         <child>
@@ -181,7 +208,7 @@
           </object>
           <packing>
             <property name="left_attach">0</property>
-            <property name="top_attach">5</property>
+            <property name="top_attach">6</property>
             <property name="width">2</property>
           </packing>
         </child>
Index: gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-tls.c
===================================================================
--- gnome-control-center-3.29.4.orig/panels/network/wireless-security/eap-method-tls.c	2018-07-17 02:32:35.000000000 +0200
+++ gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-tls.c	2018-07-24 12:14:46.053325821 +0200
@@ -139,6 +139,10 @@ add_to_size_group (EAPMethod *parent, Gt
 	g_assert (widget);
 	gtk_size_group_add_widget (group, widget);
 
+	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_tls_domain_label"));
+	g_assert (widget);
+	gtk_size_group_add_widget (group, widget);
+
 	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_tls_user_cert_label"));
 	g_assert (widget);
 	gtk_size_group_add_widget (group, widget);
@@ -181,6 +185,12 @@ fill_connection (EAPMethod *parent, NMCo
 	g_assert (widget);
 	g_object_set (s_8021x, NM_SETTING_802_1X_IDENTITY, gtk_entry_get_text (GTK_ENTRY (widget)), NULL);
 
+	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_tls_domain_entry"));
+	g_assert (widget);
+	g_object_set (s_8021x,
+	              parent->phase2 ? NM_SETTING_802_1X_PHASE2_DOMAIN_SUFFIX_MATCH : NM_SETTING_802_1X_DOMAIN_SUFFIX_MATCH,
+	              gtk_entry_get_text (GTK_ENTRY (widget)), NULL);
+
 	/* TLS private key */
 	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_tls_private_key_password_entry"));
 	g_assert (widget);
@@ -384,7 +394,7 @@ setup_filepicker (GtkBuilder *builder,
 	filter = eap_method_default_file_chooser_filter_new (privkey);
 	gtk_file_chooser_add_filter (GTK_FILE_CHOOSER (widget), filter);
 
-	/* For some reason, GTK+ calls set_current_filter (..., NULL) from 
+	/* For some reason, GTK+ calls set_current_filter (..., NULL) from
 	 * gtkfilechooserdefault.c::show_and_select_files_finished_loading() on our
 	 * dialog; so force-reset the filter to what we want it to be whenever
 	 * it gets cleared.
@@ -482,6 +492,19 @@ eap_method_tls_new (WirelessSecurity *ws
 	if (s_8021x && nm_setting_802_1x_get_identity (s_8021x))
 		gtk_entry_set_text (GTK_ENTRY (widget), nm_setting_802_1x_get_identity (s_8021x));
 
+	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_tls_domain_entry"));
+	g_assert (widget);
+	g_signal_connect (G_OBJECT (widget), "changed",
+	                  (GCallback) wireless_security_changed_cb,
+ 	                  ws_parent);
+	if (phase2) {
+		if (s_8021x && nm_setting_802_1x_get_phase2_domain_suffix_match (s_8021x))
+			gtk_entry_set_text (GTK_ENTRY (widget), nm_setting_802_1x_get_phase2_domain_suffix_match (s_8021x));
+	} else {
+		if (s_8021x && nm_setting_802_1x_get_domain_suffix_match (s_8021x))
+			gtk_entry_set_text (GTK_ENTRY (widget), nm_setting_802_1x_get_domain_suffix_match (s_8021x));
+	}
+
 	setup_filepicker (parent->builder, "eap_tls_user_cert_button",
 	                  _("Choose your personal certificate"),
 	                  ws_parent, parent, s_8021x,
@@ -531,6 +554,10 @@ eap_method_tls_new (WirelessSecurity *ws
 	if (secrets_only) {
 		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_tls_identity_entry"));
 		gtk_widget_set_sensitive (widget, FALSE);
+		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_tls_domain_label"));
+		gtk_widget_hide (widget);
+		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_tls_domain_entry"));
+		gtk_widget_hide (widget);
 		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_tls_user_cert_label"));
 		gtk_widget_hide (widget);
 		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_tls_user_cert_button"));
Index: gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-tls.ui
===================================================================
--- gnome-control-center-3.29.4.orig/panels/network/wireless-security/eap-method-tls.ui	2018-07-17 02:32:35.000000000 +0200
+++ gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-tls.ui	2018-07-24 12:14:46.053325821 +0200
@@ -43,6 +43,33 @@
           </packing>
         </child>
         <child>
+          <object class="GtkLabel" id="eap_tls_domain_label">
+            <property name="visible">True</property>
+            <property name="can_focus">False</property>
+            <property name="label" translatable="yes">_Domain:</property>
+            <property name="use_underline">True</property>
+            <property name="mnemonic_widget">eap_tls_domain_entry</property>
+            <property name="xalign">1</property>
+          </object>
+          <packing>
+            <property name="left_attach">0</property>
+            <property name="top_attach">1</property>
+          </packing>
+        </child>
+        <child>
+          <object class="GtkEntry" id="eap_tls_domain_entry">
+            <property name="visible">True</property>
+            <property name="can_focus">True</property>
+            <property name="hexpand">True</property>
+            <property name="activates_default">True</property>
+            <property name="tooltip_text" translatable="yes">Suffix of the server certificate name.</property>
+          </object>
+          <packing>
+            <property name="left_attach">1</property>
+            <property name="top_attach">1</property>
+          </packing>
+        </child>
+        <child>
           <object class="GtkLabel" id="eap_tls_user_cert_label">
             <property name="visible">True</property>
             <property name="can_focus">False</property>
@@ -53,7 +80,7 @@
           </object>
           <packing>
             <property name="left_attach">0</property>
-            <property name="top_attach">1</property>
+            <property name="top_attach">2</property>
           </packing>
         </child>
         <child>
@@ -64,7 +91,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">1</property>
+            <property name="top_attach">2</property>
           </packing>
         </child>
         <child>
@@ -78,7 +105,7 @@
           </object>
           <packing>
             <property name="left_attach">0</property>
-            <property name="top_attach">2</property>
+            <property name="top_attach">3</property>
           </packing>
         </child>
         <child>
@@ -89,7 +116,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">2</property>
+            <property name="top_attach">3</property>
           </packing>
         </child>
         <child>
@@ -104,7 +131,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">3</property>
+            <property name="top_attach">4</property>
           </packing>
         </child>
         <child>
@@ -118,7 +145,7 @@
           </object>
           <packing>
             <property name="left_attach">0</property>
-            <property name="top_attach">4</property>
+            <property name="top_attach">5</property>
           </packing>
         </child>
         <child>
@@ -129,7 +156,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">4</property>
+            <property name="top_attach">5</property>
           </packing>
         </child>
         <child>
@@ -143,7 +170,7 @@
           </object>
           <packing>
             <property name="left_attach">0</property>
-            <property name="top_attach">5</property>
+            <property name="top_attach">6</property>
           </packing>
         </child>
         <child>
@@ -156,7 +183,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">5</property>
+            <property name="top_attach">6</property>
           </packing>
         </child>
         <child>
@@ -171,7 +198,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">6</property>
+            <property name="top_attach">7</property>
           </packing>
         </child>
       </object>
Index: gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-ttls.c
===================================================================
--- gnome-control-center-3.29.4.orig/panels/network/wireless-security/eap-method-ttls.c	2018-07-17 02:32:35.000000000 +0200
+++ gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-ttls.c	2018-07-24 12:14:46.057325870 +0200
@@ -114,6 +114,10 @@ add_to_size_group (EAPMethod *parent, Gt
 	g_assert (widget);
 	gtk_size_group_add_widget (group, widget);
 
+	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_domain_label"));
+	g_assert (widget);
+	gtk_size_group_add_widget (group, widget);
+
 	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_inner_auth_label"));
 	g_assert (widget);
 	gtk_size_group_add_widget (group, widget);
@@ -154,6 +158,12 @@ fill_connection (EAPMethod *parent, NMCo
 	if (text && strlen (text))
 		g_object_set (s_8021x, NM_SETTING_802_1X_ANONYMOUS_IDENTITY, text, NULL);
 
+	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_domain_entry"));
+	g_assert (widget);
+	text = gtk_entry_get_text (GTK_ENTRY (widget));
+	if (text && strlen (text))
+		g_object_set (s_8021x, NM_SETTING_802_1X_DOMAIN_SUFFIX_MATCH, text, NULL);
+
 	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_ca_cert_button"));
 	g_assert (widget);
 	filename = gtk_file_chooser_get_filename (GTK_FILE_CHOOSER (widget));
@@ -451,6 +461,13 @@ eap_method_ttls_new (WirelessSecurity *w
 	                  (GCallback) wireless_security_changed_cb,
 	                  ws_parent);
 
+	widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_domain_entry"));
+	if (s_8021x && nm_setting_802_1x_get_domain_suffix_match (s_8021x))
+		gtk_entry_set_text (GTK_ENTRY (widget), nm_setting_802_1x_get_domain_suffix_match (s_8021x));
+	g_signal_connect (G_OBJECT (widget), "changed",
+	                  (GCallback) wireless_security_changed_cb,
+	                  ws_parent);
+
 	widget = inner_auth_combo_init (method, connection, s_8021x, secrets_only);
 	inner_auth_combo_changed_cb (widget, (gpointer) method);
 
@@ -459,6 +476,10 @@ eap_method_ttls_new (WirelessSecurity *w
 		gtk_widget_hide (widget);
 		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_anon_identity_entry"));
 		gtk_widget_hide (widget);
+		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_domain_label"));
+		gtk_widget_hide (widget);
+		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_domain_entry"));
+		gtk_widget_hide (widget);
 		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_ca_cert_label"));
 		gtk_widget_hide (widget);
 		widget = GTK_WIDGET (gtk_builder_get_object (parent->builder, "eap_ttls_ca_cert_button"));
Index: gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-ttls.ui
===================================================================
--- gnome-control-center-3.29.4.orig/panels/network/wireless-security/eap-method-ttls.ui	2018-07-17 02:32:35.000000000 +0200
+++ gnome-control-center-3.29.4/panels/network/wireless-security/eap-method-ttls.ui	2018-07-24 12:14:46.057325870 +0200
@@ -50,6 +50,33 @@
           </packing>
         </child>
         <child>
+          <object class="GtkLabel" id="eap_ttls_domain_label">
+            <property name="visible">True</property>
+            <property name="can_focus">False</property>
+            <property name="label" translatable="yes">_Domain:</property>
+            <property name="use_underline">True</property>
+            <property name="mnemonic_widget">eap_ttls_domain_entry</property>
+            <property name="xalign">1</property>
+          </object>
+          <packing>
+            <property name="left_attach">0</property>
+            <property name="top_attach">1</property>
+          </packing>
+        </child>
+        <child>
+          <object class="GtkEntry" id="eap_ttls_domain_entry">
+            <property name="visible">True</property>
+            <property name="can_focus">True</property>
+            <property name="hexpand">True</property>
+            <property name="activates_default">True</property>
+            <property name="tooltip_text" translatable="yes">Suffix of the server certificate name.</property>
+          </object>
+          <packing>
+            <property name="left_attach">1</property>
+            <property name="top_attach">1</property>
+          </packing>
+        </child>
+        <child>
           <object class="GtkLabel" id="eap_ttls_ca_cert_label">
             <property name="visible">True</property>
             <property name="can_focus">False</property>
@@ -60,7 +87,7 @@
           </object>
           <packing>
             <property name="left_attach">0</property>
-            <property name="top_attach">1</property>
+            <property name="top_attach">2</property>
           </packing>
         </child>
         <child>
@@ -71,7 +98,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">1</property>
+            <property name="top_attach">2</property>
           </packing>
         </child>
         <child>
@@ -87,7 +114,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">2</property>
+            <property name="top_attach">3</property>
           </packing>
         </child>
         <child>
@@ -101,7 +128,7 @@
           </object>
           <packing>
             <property name="left_attach">0</property>
-            <property name="top_attach">3</property>
+            <property name="top_attach">4</property>
            </packing>
          </child>
         <child>
@@ -119,7 +146,7 @@
           </object>
           <packing>
             <property name="left_attach">1</property>
-            <property name="top_attach">3</property>
+            <property name="top_attach">4</property>
           </packing>
         </child>
         <child>
@@ -134,7 +161,7 @@
           </object>
           <packing>
             <property name="left_attach">0</property>
-            <property name="top_attach">4</property>
+            <property name="top_attach">5</property>
             <property name="width">2</property>
           </packing>
         </child>
